skbio/diversity/_driver.py

```diff
@@ -180,9 +182,114 @@ def alpha_diversity(metric, counts, ids=None, validate=True, **kwargs):
     return pd.Series(results, index=ids)
 
 
+@deprecated(as_of='0.4.2-dev', until='0.5.0',
+            reason=('The return type is unstable. Developer caution is '
+                    'advised. The resulting DistanceMatrix object may include '
+                    'zeros when distance has not been calculated, and '
+                    'therefore can be misleading.'))
+def partial_beta_diversity(metric, counts, ids=None, validate=None,
+                           id_pairs=None, **kwargs):
+    """Compute distances only between specified ID pairs
+
+    Parameters
+    ----------
+    metric : callable
+        The pairwise distance function to apply. ``metric`` must be resolvable
+        by scikit-bio (e.g., UniFrac methods), or must be callable.
+    counts : 2D array_like of ints or floats
+        Matrix containing count/abundance data where each row contains counts
+        of OTUs in a given sample.
+    ids : iterable of strs
+        Identifiers for each sample in ``counts``.
+    id_pairs : iterable of tuple, optional
+        An iterable of tuples of IDs to compare (e.g., ``[('a', 'b'), ('a',
+        'c'), ...])``. If specified, the set of IDs described must be a subset
+        of ``ids``.
+    validate : ignored
+        This kwarg is preserved to maintain consistency with the beta_diversity
+        signature.
+    kwargs : kwargs, optional
+        Metric-specific parameters.
+
+    Returns
+    -------
+    skbio.DistanceMatrix
+        Distances between all pairs of samples (i.e., rows). The number of
+        rows and columns will be equal to the number of rows in ``counts``.
+
+        Warning: zeros in this matrix may represent a zero distance, or may
+        represent a pairwise distance that was not calculated.
+
+    Raises
+    ------
+    ValueError
+        If ``ids`` are not specified.
+        If ``id_pairs`` are not a subset of ``ids``.
+        If ``metric`` is not a callable.
+        If duplicates are observed in ``id_pairs``.
+
+    See Also
+    --------
+    skbio.beta_diversity
+    skbio.diversity.get_beta_diversity_metrics
+
+    """
+    if ids is None:
+        raise ValueError("`ids` must be specified")
+
+    if id_pairs is None:
+        # determine all pairs of the upper triangle.
+        # alternatively, could just call beta_diversity?
+        id_pairs = []
+        for row, i in enumerate(ids):
+            for j in ids[row+1:]:
+                id_pairs.append((i, j))
+
+    all_ids_in_pairs = set(itertools.chain.from_iterable(id_pairs))
```

Comment:
The current implementation will fail if providing an iterable because the code takes multiple passes through `id_pairs`. Can you add a test case for an iterable `id_pairs`? e.g., call `iter([('a','b'), ('c', 'd')])` or pass a generator